ai: add configurable item drop behaviour

lahm86 opened the pull request:
Resolves #967.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This follows on from #1000, and allows all but a few enemies to carry and drop items for Lara to collect. The hardcoded drops have been removed and the gameflow is now responsible for the setup. The default gameflow is configured in line with OG behaviour/requirements.

`convert_dropped_guns` in the gameflow controls whether guns are converted to ammo when dropped and Lara already has that gun. This is `false` by default to avoid extra ammo pickups when killing Pierre, Cowboy and Kold in the original levels.

Enemies that are *not* capable of carrying and dropping are:
- `O_CROCODILE` `O_ALLIGATOR` `O_RAT` `O_VOLE`  - having items spawn in the water would be difficult to track with changing water levels I feel, and having floating pickups would not be in-keeping with normal experience.
- `O_WARRIOR1` `O_WARRIOR2` `O_WARRIOR3` - these explode in mid-air and don't follow the standard `Creature_Animate` routine on death. We could manually spawn the items in each object's control, and work out the floor height, but the item would appear out of nowhere if the creature were high above land when killed, so again it feels non-standard.

Pods and centaur statues can be referenced as carriers in the gameflow; the system will give the items to the enemies that actually spawn, provided those are suitable carriers (empty pods are excluded).

The statistics take into account carried items. I have added a function here for common behaviour when testing triggered killable items.

In terms of saving and loading, items will respawn accordingly if Lara hasn't picked them up yet and equally they will not if she has done so. For testing, I used legacy ATI saves, V3 of the TR1X save format (2.16) and the current V4. In each case, I tested Pierre, Cowboy, Skaterboy and Kold in six scenarios - the following with `convert_dropped_guns` set to `false` and then again with it set to `true`.
- Before killing the enemy
- After killing the enemy, but before collecting their item(s)
- After killing the enemy, and after collecting their item(s)

I also used a test level to ensure items allocated to every potential carrier type spawn as expected, and that items allocated to non-carriers are ignored.

I will follow-up this PR with a document update to create a gameflow MD file, so that builders have a go-to reference for this feature in detail, as well as other gameflow aspects.

lahm86 committed:
ai: add configurable item drop behaviour

This replaces all hardcoded enemy item drops with a flexible system
using the gameflow. Specific enemies can carry pickup items, which
spawn on death and are remembered between saves. Dropped guns can also
optionally be converted to ammo if Lara has the gun already.
Support is also included for Pierre, Cowboy, Skaterboy and Kold to
handle their items accordingly when using legacy save files.

Resolves #967.

rr- commented on src/game/carrier.c:
```diff
@@ -0,0 +1,213 @@
+            }
+        }
+    }
+}
```
Hi, let's add more comments in the functions what's their purpose.

rr- commented on src/game/carrier.c:
```diff
@@ -0,0 +1,213 @@
+#include <stdbool.h>
+#include <stddef.h>
+
+#define AMMO_OFFSET 4
```
Can we do without this const?

lahm86 replied:
Perhaps an explicit switch statement to convert gun to ammo would be better, or a map even.

rr- replied:
Yes that's exactly what I had on my mind.

rr- commented on src/game/carrier.c:
```diff
@@ -0,0 +1,213 @@
+    ITEM_INFO *item = &g_Items[item_num];
+    if (Carrier_IsObjectType(item->object_number, m_PlaceholderObjects)) {
+        int16_t child_item_num = *(int16_t *)item->data;
+        item = &g_Items[child_item_num];
```
It's amazing that you thought of this even.

rr- commented on src/game/carrier.c:
```diff
@@ -0,0 +1,213 @@
+        }
+    }
+    return false;
+}
```
Can the `int16_t` throughout this file become `GAME_OBJECT_ID`?

lahm86 replied:
Yes, that would be much neater I think.

lahm86 committed:
ai: reconfigure item drop handling

Converted int16_t to GAME_OBJECT_ID in the carrier module; changed the
approach for gun-to-ammo conversion and legacy enemy testing; changed
invalid drop object ID handling - these will no longer invalidate the
entire array for that enemy.
Added comments to carrier functions for clarity.

Part of #967.

walkawayy reviewed (approved):
Really nice!
